src/main/plugins/org.dita.base/xsl/preprocess/conrefImpl.xsl

```diff
@@ -898,4 +895,37 @@ See the accompanying LICENSE file for applicable license.
     </xsl:call-template>
   </xsl:template>
 
+  <xsl:template match="*[@conref:orig-id]" mode="clean">
+    <xsl:copy>
+      <xsl:copy-of select="@* except @conref:orig-id"/>
+      <xsl:variable name="orig-id" select="@conref:orig-id" as="attribute()"/>
+      <xsl:variable name="topic" select="ancestor::*[contains(@class, ' topic/topic ')][1]" as="element()"/>
+      <xsl:variable name="content" select="$topic/*[not(contains(@class, ' topic/topic '))]" as="element()*"/>
+      <xsl:variable name="first" select="($content//@conref:orig-id[. = $orig-id])[1]" as="attribute()"/>
+      <xsl:variable name="ids" as="element()*">
+        <xsl:for-each select="$content">
+          <xsl:sequence select="key('id', $orig-id, .)"/>
+        </xsl:for-each>
+      </xsl:variable>
+      <xsl:choose>
+        <xsl:when test="empty($ids) and $orig-id is $first">
+          <xsl:attribute name="id" select="$orig-id"/>
+        </xsl:when>
+        <xsl:otherwise>
+          <xsl:message>
+            <xsl:processing-instruction name="level" select="'DEBUG'"/>
+            <xsl:value-of>Rewrite conref content @id from "<xsl:value-of select="@conref:orig-id"/>" to "<xsl:value-of select="@id"/>"</xsl:value-of>
```

Comment:
@robander @infotexture Should we have ID for every message like this? We don't really need it for localization of the message text, but it would make it easier to find code that threw the message, be it error or debug.

Reply:
Looking at XSLT code, we actually have only few cases where we throw ad hoc messages like this. This should be moved to messages.xml and given a message ID.